### DISCUSSION TITLE
TorchVision layer freeze and TIMM support

### DISCUSSION THREAD

**Discussion by someuser2026** [ultralytics/ultralytics#21464] [Q&A]
Hi, Appreicate the torchvision addition to the library. It is amazing! I have two seperate queries:

1. Is it possible to freeze indivisual layers or stages inside the torchvision network used using the TorchVision module? For example if I use a convnext backone and want to freeze the first stage is it possible?
2. Is there integration for timm pretrained models also, since it is quite similar in syntax to the torch vision models?

Thank you. Greatly appreciate your time.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @someuser2026, thank you for your interest in Ultralytics and for the positive feedback on the TorchVision integration! 🚀 This is an automated response to help get you started while an Ultralytics engineer reviews your questions.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

If your queries are for custom training or feature requests, providing more details—such as code snippets or specific use cases—will help us understand your needs better. Be sure to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for guidance.

For more information, check out the [Docs](https://docs.ultralytics.com/) where you’ll find detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

Join our community for further discussion or support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth threads
- [Subreddit](https://www.reddit.com/r/Ultralytics/) to connect with other users

## Upgrade

To ensure you have the latest features and fixes, upgrade the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs in these up-to-date verified environments (dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

An Ultralytics engineer will assist you soon. Thank you for contributing to the project!

**Reply by glenn-jocher** [reply]
Great questions! For layer freezing in TorchVision backbones, you can access the underlying model via the TorchVision module and freeze specific layers by setting `requires_grad=False` on their parameters. For example, after model creation you could iterate through the backbone layers and freeze the first stage parameters.

Regarding TIMM integration, there isn't direct support for TIMM models in the same way as TorchVision currently. However, this would be a valuable feature addition - you could consider opening a feature request issue to discuss potential TIMM integration with the development team.
